File: authentication/src/main/java/authentication/config/aws/DataSourceConfig.java
package authentication.config.aws;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import javax.sql.DataSource;

@Configuration
@Profile("prod")
@RequiredArgsConstructor
public class DataSourceConfig {

    private final DbSecretLoader dbSecretLoader;

    @Bean
    public DataSource dataSource() throws JsonProcessingException {
        DbSecretProperties properties = dbSecretLoader.load("prod/db/parameter");

        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(properties.getUrl());
        config.setUsername(properties.getUsername());
        config.setPassword(properties.getPassword());
        config.setDriverClassName("com.mysql.cj.jdbc.Driver");

        return new HikariDataSource(config);
    }
}

File: api/src/main/kotlin/api/config/DataSourceConfig.kt
package api.config

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider
import javax.sql.DataSource

@Configuration
@Profile("prod")
class DataSourceConfig(
    private val secretsManagerConfig: SecretsManagerConfig
) {

    @Bean
    fun dataSource(): DataSource {
        val secrets = secretsManagerConfig.getRdsSecrets()
        val jdbcUrl = secrets["url"]!!
        val username = secrets["username"]!!
        val password = secrets["password"]!!

        val config = HikariConfig().apply {
            this.jdbcUrl = jdbcUrl
            this.username = username
            this.password = password
        }

        return HikariDataSource(config)
    }
}
